extensions/vertx/runtime/src/main/java/io/quarkus/vertx/VertxContextSupport.java

```diff
@@ -0,0 +1,98 @@
+package io.quarkus.vertx;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.ManagedContext;
+import io.quarkus.vertx.core.runtime.VertxCoreRecorder;
+import io.quarkus.vertx.core.runtime.context.VertxContextSafetyToggle;
+import io.smallrye.common.vertx.VertxContext;
+import io.smallrye.mutiny.Uni;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+
+/**
+ * Provides utility methods to work with Vertx duplicated context.
+ *
+ * @see VertxContext
+ */
+public final class VertxContextSupport {
+
+    private VertxContextSupport() {
+    }
+
+    /**
+     * Subscribes to the supplied {@link Uni} on a Vertx duplicated context; blocks the current thread and waits for the result.
+     * <p>
+     * If it's necessary, the CDI request context is activated during execution of the asynchronous code.
+     *
+     * @param uniSupplier
+     * @throws IllegalStateException If called on an event loop thread.
+     */
+    public static <T> T subscribeAndBlock(Supplier<Uni<T>> uniSupplier) throws Throwable {
+        T ret = null;
+
+        Context context = Vertx.currentContext();
+        if (context == null) {
+            Vertx vertx = VertxCoreRecorder.getVertx().get();
+            context = VertxContext.getOrCreateDuplicatedContext(vertx);
+        } else {
+            // Executed on a vertx thread...
+            if (Context.isOnEventLoopThread()) {
+                throw new IllegalStateException("VertxContextSupport#subscribeAndBlock() must not be called on an event loop!");
+            }
+            context = VertxContext.getOrCreateDuplicatedContext(context);
+        }
+
+        VertxContextSafetyToggle.setContextSafe(context, true);
+
+        CompletableFuture<T> cf = new CompletableFuture<>();
```

Comment:
We can simplify it:

```
return Uni.createFrom().emitter(e -> {
   context.runOnContext(() -> {
      try {
            // .... 
            emitter.complete(ret);
      }  catch(Throwable t) {
             emitter.failed(t);
      }
   }).onTerminate().invoke(() -> terminate.run())
}).await().indefinitely();
```

Or something like this.

Reply:
I can't say that I understand how this would simplify things ;-). It's one more lambda vs one `CompletableFuture`. We need to subscribe to the supplied `Uni` and call `emitter.complete()/failed()` anyway, or am I missing something? Also where does the `onTerminate()` come from? Is it the `Uni.onTermination()`?

That said, I do understand that it's more Mutiny-idiomatic so I'll try to do my best... :shrug:

Reply:
@cescoffier how about now?